Possibility to report violations of the Content-Security-Policy

mfoXen opened the pull request:
Solving issue #0019307 to possibly find reason on issue #0017491.
First added some source-comments.
Second splitted the construction of this csp-header to simplyfie future changes.
Third added an option to enable/disable an report-uri-directiv to the csp-header.
Finally added \csp\csp_report.php which do log any violations.

To get such a report set $g_report_csp_violation = ON and start using mantis. Possibly you have to create a subfolder "logs" to "csp"-folder.
As far as an violation appears, the corresponding logfile will be extended - and this way we'll get noticed which csp-directive results in blocking those gravatars.

I wasn't sure where to put csp_report*.php and their logs, that's why simply created new subfolder. I'm sure you would relocate that...

mfoXen committed:
adding some comments to func http_security_headers

mfoXen committed:
Constructing CSP-Header directive by directive

mfoXen committed:
Added and implemented option to report CPS-violations, solving issue #0019307

dregad commented:
Thanks for your contribution. I have similar code on a local branch which I rebase whenever I need to troubleshoot CSP issues... This is much more convenient, and it kind of makes sense to have this option for Mantis/plugin developers.

I'll review this as time allows.

rombert commented:
When reviewing please consider that csp/csp_report.php left as-is is at least a DOS vulnerability which allows an attacker to fill up a disk by just posting large JSON payloads.

mfoXen commented:
Thx for your responses. Well, I'm not very familiar with JSON and the payloads, but appreciate any help as well as positive modifications.
May it helpfull, to move all this into the admin-folder? As this option should not be enabled (permanently) in a production environment it could be delete as well, right?
Although writing to file should only be done if ($data = json_decode($data)) - I'll change this.

mfoXen committed:
code-formating

added spaces after opening braces and before closing braces

mfoXen committed:
changed to unix-style linebreaks

mfoXen committed:
typo in comment

mfoXen committed:
Validating JSON-input before logging

Only write to log, if we got a valid CSP-report that is only a single key called csp-report and subkeys as defined bei CSP v1 [http://www.w3.org/TR/2012/CR-CSP-20121115/]; thx for hint @rombert

mfoXen commented:
Well, I know implemented some checks on the JSON input. Hopefully this is much more secure? @rombert 
Btw I also changed to mantis-default-linebreaks which is unix. :)

mfoXen committed:
fix: json_decode to array

rombert commented:
@mfoXen - sorry, I didn't find the time to review this more in depth - maybe @dregad will be able to suggest a better way of securing the page.

As it stands, it still allows anyone with a _valid_ payload to DOS the server by continually uploading valid payloads

l266509 commented:
@vboctor @giallu @daryn @rombert @atrol @grangeway @dregad @Dentxinho @bieli @tomcarlson @syncguru @diegovice @caribe @fager @nikoutel @6RUN0 @AgnidykhBayky @sayyidsh @PixelstreamAU @JohnMS @mikeching @g-taki @benv8nam @stew721 @sayazyi @lezouretalice @SuriyaaKudoIsc @solovets @kacmem @xiaojiangang @chenyongze @myu9882 @markmarco16 @mgretler @lincolnwung @krime @scottmatheina @FavioGalvis @vand3rson @like9729 @badfiles @zenithtekla @SureShinde @netsf @csatwood @miao606 @mdj20130907 @bitlin @ioroid @supertassu @TasCraft @fernandodelucca @ulkuderner @vbathke @hersonpc @jrtorres42 @thejuanwhocode @cyuankai @mantianxingxing @shirleycharlin @boby75 @Lordshell @BurningWind @jaguarxii @Praymundo @johnzastrow @ewmartinez @marcioklein @lypzln @nospam1961 @jllano @lincoln56 @modir @xiaocnie @sivpro @planetbrett @teodortalov @ltpgt @pbcreativesolutions @Gantha @Gerolkae @gunsmoker @nikerabbit @translatewiki  Whats status of this PR?
